pubpub/pubpub
# Reference shortcut and suggestions dropdown

This PR adds support for inline referencing of nodes in the editor based on a trigger character `@`. When the cursor is present within a block of text preceded by `@`, a dropdown will appear showing reference-able nodes whose labels contain the text following the trigger character, e.g:

<img width="200" alt="Screen Shot 2020-09-30 at 1 39 25 PM" src="https://user-images.githubusercontent.com/6402908/94720353-5da81500-0322-11eb-94e5-be284c8fa08a.png">

The state of the query within ProseMirror and React is stored in a new class called `SuggestionManager` which tracks the current highlighted index, filtered nodes, and plugin state used to replace the query with a true reference (called `params`).

Here's a gif of the shortcut and suggestions dropdown in action:

![captured (1)](https://user-images.githubusercontent.com/6402908/94722060-c2646f00-0324-11eb-9a4c-66f0edefd9ce.gif)

_Test Plan_

1. Create a Pub with multiple, reference-able nodes (e.g. media).
2. Type within an existing text paragraph, or new paragraph, starting with`@`. You should see a list of all reference-able nodes in the document.
3. Type a partial label for a node. The list of references should narrow based on the text after the `@` symbol.
4. Use the up and down arrows to navigate the menu.
5. Use either the enter key or the mouse button to select a node.
6. The shortcut and query (e.g `@fig`) should be replaced with a reference.
